In [1]:
import numpy as np
import pandas as pd
from sklearn import datasets, model_selection
from sklearn.datasets import load_digits, load_breast_cancer
from sklearn.naive_bayes import BernoulliNB, GaussianNB, MultinomialNB

Смотрим на датасет

In [2]:
digits = load_digits()
df1 = pd.DataFrame(data=np.c_[digits.data, digits.target])
df1.rename(columns={64:'target'}, inplace=True)
df1.head()

0    1    2     3     4     5    6    7    8    9   ...     55   56   57  \
0  0.0  0.0  5.0  13.0   9.0   1.0  0.0  0.0  0.0  0.0   ...    0.0  0.0  0.0   
1  0.0  0.0  0.0  12.0  13.0   5.0  0.0  0.0  0.0  0.0   ...    0.0  0.0  0.0   
2  0.0  0.0  0.0   4.0  15.0  12.0  0.0  0.0  0.0  0.0   ...    0.0  0.0  0.0   
3  0.0  0.0  7.0  15.0  13.0   1.0  0.0  0.0  0.0  8.0   ...    0.0  0.0  0.0   
4  0.0  0.0  0.0   1.0  11.0   0.0  0.0  0.0  0.0  0.0   ...    0.0  0.0  0.0   

    58    59    60    61   62   63  target  
0  6.0  13.0  10.0   0.0  0.0  0.0     0.0  
1  0.0  11.0  16.0  10.0  0.0  0.0     1.0  
2  0.0   3.0  11.0  16.0  9.0  0.0     2.0  
3  7.0  13.0  13.0   9.0  0.0  0.0     3.0  
4  0.0   2.0  16.0   4.0  0.0  0.0     4.0  

[5 rows x 65 columns]

In [3]:
cancer = load_breast_cancer()
df2 = pd.DataFrame(data=np.c_[cancer.data, cancer.target])
df2.rename(columns={30:'target'}, inplace=True)
df2.head()

0      1       2       3        4        5       6        7       8  \
0  17.99  10.38  122.80  1001.0  0.11840  0.27760  0.3001  0.14710  0.2419   
1  20.57  17.77  132.90  1326.0  0.08474  0.07864  0.0869  0.07017  0.1812   
2  19.69  21.25  130.00  1203.0  0.10960  0.15990  0.1974  0.12790  0.2069   
3  11.42  20.38   77.58   386.1  0.14250  0.28390  0.2414  0.10520  0.2597   
4  20.29  14.34  135.10  1297.0  0.10030  0.13280  0.1980  0.10430  0.1809   

         9   ...       21      22      23      24      25      26      27  \
0  0.07871   ...    17.33  184.60  2019.0  0.1622  0.6656  0.7119  0.2654   
1  0.05667   ...    23.41  158.80  1956.0  0.1238  0.1866  0.2416  0.1860   
2  0.05999   ...    25.53  152.50  1709.0  0.1444  0.4245  0.4504  0.2430   
3  0.09744   ...    26.50   98.87   567.7  0.2098  0.8663  0.6869  0.2575   
4  0.05883   ...    16.67  152.20  1575.0  0.1374  0.2050  0.4000  0.1625   

       28       29  target  
0  0.4601  0.11890     0.0  
1  0.2750  0.08902     0.0  
2  0.3613  0.08758     0.0  
3  0.6638  0.17300     0.0  
4  0.2364  0.07678     0.0  

[5 rows x 31 columns]

Видим, что признаки в первом датасете - целые числа, тогда как во втором - дробные.

Отделяем матрицу фичей от таргета

In [4]:
X1, y1 = load_digits(return_X_y=True)
X2, y2 = load_breast_cancer(return_X_y=True)

Создаем классификаторы

In [5]:
bernoulli = BernoulliNB()
multinomial = MultinomialNB()
gaussian = GaussianNB()

Обучаем на первом и втором датасетах, смотрим результат

In [6]:
print model_selection.cross_val_score(bernoulli,   X1, y1).mean(), \
      model_selection.cross_val_score(multinomial, X1, y1).mean(), \
      model_selection.cross_val_score(gaussian,    X1, y1).mean()

0.825823650778 0.870877148974 0.818600380355


In [7]:
print model_selection.cross_val_score(bernoulli,   X2, y2).mean(), \
      model_selection.cross_val_score(multinomial, X2, y2).mean(), \
      model_selection.cross_val_score(gaussian,    X2, y2).mean()

0.627420402859 0.894579040193 0.936749280609


# Результаты

1. Максимальное качество классификации на датасете breast_cancer - 0.936749280609.
2. Максимальное качество классификации на датасете digits - 0.870877148974.
3. Из приведенных верны следующие утверждения:

(c) Мультиномиальное распределение лучше показало себя на выборке с целыми неотрицательными значениями признаков

(d) На вещественных признаках лучше всего сработало нормальное распределение